### PR TITLE
Add term simplification to Python bindings

### PR DESCRIPTION
**Pull request by Baltoli**
This PR exposes the concrete term simplification routines already used by the booster to the Python bindings so that they can be called from Pyk. Specifically, it adds two free functions to the `runtime` module:
```python
simplify_pattern(Pattern, Sort) -> Pattern
simplify_bool_pattern(Pattern) -> bool
```

These are exposed as free functions because they depend on the runtime library and so can't be automatically bound to the `Pattern` class as methods; I suggest Pyk probably wants to monkey-patch them into its `Pattern` class along with some kind of dynamic check for whether the `runtime` module has been loaded or not.

Previously, the C and Python bindings didn't share a great deal of code; the simplification routine is the first bit of non-trivial behaviour that they need to share. I have therefore started a reorganisation of the bindings code to extract as much functionality to a common library, which has the additional benefit of reducing the amount of memory-leak prone handling of C wrapper structures. To keep the diff in this PR reasonable I have not completed this migration, but this is future work that I will complete in a future PR.

A summary of the changes in this PR is as follows:
* Build a common bindings library that the C and Python bindings link against.
* Extract the minimal set of features from the existing C bindings that the Python bindings require to perform term simplification, and port the C bindings over to calling the extracted common features.
* Add a test case to the Python bindings test suite that performs a few small simplifications of sort `Int` and `Bool`; these are really just smoke tests and Pyk should test the feature more thoroughly.

### CODE REVIEW

**Comment by Baltoli** [comment]
I have also tested the booster's integration tests with this change applied and there is no breakage to the C bindings.

**Review by tothtamas28 — COMMENTED** [review]
Looks good in terms of functionality.

Python tests: https://github.com/runtimeverification/pyk/pull/736

**Review by dwightguth — CHANGES_REQUESTED** [review]
It's very difficult to follow these changes because they are a mix of a refactoring that moves a lot of code around, and new features. I'm concerned that despite my best efforts, I still wasn't really able to determine what code was "new" and review it effectively.

I think you need to refactor the history of this PR to more obviously break it into two components: the moving around of existing code, and the addition of new code.

File: bindings/c/lib.cpp [hunk 1]
```diff
@@ -185,14 +186,8 @@ kore_pattern *kore_pattern_new_token_with_len(
...
-  auto inj_sym = kore_symbol_new("inj");
-  kore_symbol_add_formal_argument(inj_sym, from);
-  kore_symbol_add_formal_argument(inj_sym, to);
-
-  auto inj = kore_composite_pattern_from_symbol(inj_sym);
-  kore_composite_pattern_add_argument(inj, term);
-
-  kore_symbol_free(inj_sym);
+  auto inj = new kore_pattern;
+  inj->ptr_ = kllvm::bindings::make_injection(term->ptr_, from->ptr_, to->ptr_);
```
Comment (Baltoli) [inline bindings/c/lib.cpp:189]: @dwightguth As an example of how to look at the changes in this diff, the `kllvm::bindings::make_injection` function contains code that is _equivalent_ to the C on the LHS here, but using the underlying C++ structures directly. All the changes in this commit are of that form.

File: bindings/core/src/core.cpp [hunk 2]
```diff
@@ -1 +1,83 @@
...
+    std::shared_ptr<KORESort> to) {
+  auto inj_sym = KORESymbol::Create("inj");
+
+  inj_sym->addFormalArgument(from);
+  inj_sym->addFormalArgument(to);
+
+  auto inj = KORECompositePattern::Create(std::move(inj_sym));
+  inj->addArgument(term);
+
+  return inj;
```
Comment (Baltoli) [inline bindings/core/src/core.cpp:26]: The restated C++ code from my other comment is here.

**Review by dwightguth — APPROVED** [review]

**Pull request merged**